### ISSUE TITLE
readtags: strange error message "cannot open tag file: no error" (transferred from citre by @masatake)

### ISSUE THREAD

**Issue by lambertlulala** [universal-ctags/ctags#3370]
# Description
Hello maintainers!

I use `citre-peek` for Linux kernel code on Windows, and get the exception like this:
![image](https://user-images.githubusercontent.com/76695321/165936665-9d632b7d-6326-4a7a-a4fc-83309e575c20.png)
The symbol to be searched is `WARN_ON_ONCE` in `d:\git\linux-master_modify\linux-master\kernel\cpu.c`.

And this is another error when I use citre-peek to search `spin_lock` in `d:\git\linux-master_modify\linux-master\ipc\mqueue.c`.
![image](https://user-images.githubusercontent.com/76695321/165940604-ba80dd5f-2f6e-4f09-8d36-1df0072d1191.png)

But everything is fine when I use citre-peek on other projects.
Note the tag file is in `d:\git\linux-master_modify\linux-master\.tags\.!.tags`.

# Environment
- OS: Windows 10
- Emacs version: GNU Emacs 28.1 (build 2, x86_64-w64-mingw32) of 2022-04-22
- citre
   - Version: 20220427.1203
   - Commit: 87e2cbf3b2ae6d59ec919a2dcb38e56ccfa5ec14
- ctags version
```
Universal Ctags 5.9.0(p5.9.20220424.0), Copyright (C) 2015-2022 Universal Ctags Team
Universal Ctags is derived from Exuberant Ctags.
Exuberant Ctags 5.8, Copyright (C) 1996-2009 Darren Hiebert
  Compiled: Apr 28 2022, 11:40:13
  URL: https://ctags.io/
  Optional compiled features: +win32, +wildcards, +regex, +gnulib_regex, +internal-sort, +unix-path-separator, +iconv, +option-directory, +xpath, +json, +interactive, +yaml, +case-insensitive-filenames, +packcc, +optscript, +pcre2
```

How do I resolve this issue?

**Comment by lambertlulala** [comment]
Thank you for your summarization and help.

Yes, `od` is available on my platform. Sure, if text content is more preferable.

**Comment by masatake** [comment]
Before starting bug hunting I would like to verify the condition and status of the targets and environment.

```
$ ls -l original.tags shorten.tags
$ wc -l original.tags shorten.tags
```
Can I see the output of the above two commands?

**Comment by lambertlulala** [comment]
Sure.

$ ls -l original.tags shorten.tags
```
-rw-r--r-- 1  2331246140 Apr 29 13:51 original.tags
-rw-r--r-- 1       17887      Apr 29 22:29 shorten.tags
```

$ wc -l original.tags shorten.tags
```
  10786857 original.tags
       100 shorten.tags
  10786957 total
```

**Comment by masatake** [comment]
Thank you. The biggest question is why readtags works well with shorten.tags but doesn't work well with original.tags. 

As you may know already, a (sorted) tags file has some entries starting from `!_` at the beginning. They are called pseudo tags.
Pseudo tags provide meta-information about the tags file and they are very related to the open operation of readtags.
I guess there are subtle differences between the pseudo tags of the original.tags and that of the shorten.tags.

Could you show the output of the following commands:

```
$ readtags -t original.tags - '$$divU_6'
$ readtags -t shorten.tags - '$$divU_6'
$ od -A x -t x1z -v original.tags | head -60
$ od -A x -t x1z -v shorten.tags | head -60
```

**Comment by lambertlulala** [comment]
> As you may know already, a (sorted) tags file has some entries starting from `!_` at the beginning. They are called pseudo tags. Pseudo tags provide meta-information about the tags file and they are very related to the open operation of readtags. I guess there are subtle differences between the pseudo tags of the original.tags and that of the shorten.tags.

That's cool, I read some description about citre, and it states something about meta data, after your explanation I got it.

@masatake 
```
$ readtags -t original.tags - '$$divU_6'
C:\msys64\mingw64\bin\readtags.exe: cannot open tag file: no error: original.tags

$ readtags -t shorten.tags - '$$divU_6'
$$divU_6        arch/parisc/kernel/parisc_ksyms.c       /^extern void $$divU_6(void);$/

$ od -A x -t x1z -v original.tags | head -60
000000 21 5f 43 49 54 52 45 5f 43 4d 44 09 63 74 61 67  >!_CITRE_CMD.ctag<
000010 73 7c 2d 6f 7c 25 54 41 47 53 46 49 4c 45 25 7c  >s|-o|%TAGSFILE%|<
000020 2d 2d 6c 61 6e 67 75 61 67 65 73 3d 43 2c 43 2b  >--languages=C,C+<
000030 2b 7c 2d 2d 6b 69 6e 64 73 2d 61 6c 6c 3d 2a 7c  >+|--kinds-all=*|<
000040 2d 2d 66 69 65 6c 64 73 3d 2a 7c 2d 2d 65 78 74  >--fields=*|--ext<
000050 72 61 73 3d 2a 7c 2d 52 09 2f 63 6f 6d 6d 61 6e  >ras=*|-R./comman<
000060 64 20 6c 69 6e 65 20 74 6f 20 67 65 6e 65 72 61  >d line to genera<
000070 74 65 20 74 68 69 73 20 74 61 67 73 20 66 69 6c  >te this tags fil<
000080 65 2f 3b 22 0d 0a 21 5f 54 41 47 5f 46 49 4c 45  >e/;"..!_TAG_FILE<
000090 5f 46 4f 52 4d 41 54 09 32 09 2f 65 78 74 65 6e  >_FORMAT.2./exten<
0000a0 64 65 64 20 66 6f 72 6d 61 74 3b 20 2d 2d 66 6f  >ded format; --fo<
0000b0 72 6d 61 74 3d 31 20 77 69 6c 6c 20 6e 6f 74 20  >rmat=1 will not <
0000c0 61 70 70 65 6e 64 20 3b 22 20 74 6f 20 6c 69 6e  >append ;" to lin<
0000d0 65 73 2f 3b 22 09 65 78 74 72 61 73 3a 70 73 65  >es/;".extras:pse<
0000e0 75 64 6f 0d 0a 21 5f 54 41 47 5f 46 49 4c 45 5f  >udo..!_TAG_FILE_<
0000f0 53 4f 52 54 45 44 09 31 09 2f 30 3d 75 6e 73 6f  >SORTED.1./0=unso<
000100 72 74 65 64 2c 20 31 3d 73 6f 72 74 65 64 2c 20  >rted, 1=sorted, <
000110 32 3d 66 6f 6c 64 63 61 73 65 2f 3b 22 09 65 78  >2=foldcase/;".ex<
000120 74 72 61 73 3a 70 73 65 75 64 6f 0d 0a 21 5f 54  >tras:pseudo..!_T<
000130 41 47 5f 4f 55 54 50 55 54 5f 45 58 43 4d 44 09  >AG_OUTPUT_EXCMD.<
000140 6d 69 78 65 64 09 2f 6e 75 6d 62 65 72 2c 20 70  >mixed./number, p<
000150 61 74 74 65 72 6e 2c 20 6d 69 78 65 64 2c 20 6f  >attern, mixed, o<
000160 72 20 63 6f 6d 62 69 6e 65 56 32 2f 3b 22 09 65  >r combineV2/;".e<
000170 78 74 72 61 73 3a 70 73 65 75 64 6f 0d 0a 21 5f  >xtras:pseudo..!_<
000180 54 41 47 5f 4f 55 54 50 55 54 5f 46 49 4c 45 53  >TAG_OUTPUT_FILES<
000190 45 50 09 73 6c 61 73 68 09 2f 73 6c 61 73 68 20  >EP.slash./slash <
0001a0 6f 72 20 62 61 63 6b 73 6c 61 73 68 2f 3b 22 09  >or backslash/;".<
0001b0 65 78 74 72 61 73 3a 70 73 65 75 64 6f 0d 0a 21  >extras:pseudo..!<
0001c0 5f 54 41 47 5f 4f 55 54 50 55 54 5f 4d 4f 44 45  >_TAG_OUTPUT_MODE<
0001d0 09 75 2d 63 74 61 67 73 09 2f 75 2d 63 74 61 67  >.u-ctags./u-ctag<
0001e0 73 20 6f 72 20 65 2d 63 74 61 67 73 2f 3b 22 09  >s or e-ctags/;".<
0001f0 65 78 74 72 61 73 3a 70 73 65 75 64 6f 0d 0a 21  >extras:pseudo..!<
000200 5f 54 41 47 5f 50 41 54 54 45 52 4e 5f 4c 45 4e  >_TAG_PATTERN_LEN<
000210 47 54 48 5f 4c 49 4d 49 54 09 39 36 09 2f 30 20  >GTH_LIMIT.96./0 <
000220 66 6f 72 20 6e 6f 20 6c 69 6d 69 74 2f 3b 22 09  >for no limit/;".<
000230 65 78 74 72 61 73 3a 70 73 65 75 64 6f 0d 0a 21  >extras:pseudo..!<
000240 5f 54 41 47 5f 50 52 4f 43 5f 43 57 44 09 44 3a  >_TAG_PROC_CWD.D:<
000250 2f 67 69 74 2f 6c 69 6e 75 78 2d 6d 61 73 74 65  >/git/linux-maste<
000260 72 5f 6d 6f 64 69 66 79 2f 6c 69 6e 75 78 2d 6d  >r_modify/linux-m<
000270 61 73 74 65 72 2f 09 2f 2f 3b 22 09 65 78 74 72  >aster/.//;".extr<
000280 61 73 3a 70 73 65 75 64 6f 0d 0a 21 5f 54 41 47  >as:pseudo..!_TAG<
000290 5f 50 52 4f 47 52 41 4d 5f 41 55 54 48 4f 52 09  >_PROGRAM_AUTHOR.<
0002a0 55 6e 69 76 65 72 73 61 6c 20 43 74 61 67 73 20  >Universal Ctags <
0002b0 54 65 61 6d 09 2f 2f 3b 22 09 65 78 74 72 61 73  >Team.//;".extras<
0002c0 3a 70 73 65 75 64 6f 0d 0a 21 5f 54 41 47 5f 50  >:pseudo..!_TAG_P<
0002d0 52 4f 47 52 41 4d 5f 4e 41 4d 45 09 55 6e 69 76  >ROGRAM_NAME.Univ<
0002e0 65 72 73 61 6c 20 43 74 61 67 73 09 2f 44 65 72  >ersal Ctags./Der<
0002f0 69 76 65 64 20 66 72 6f 6d 20 45 78 75 62 65 72  >ived from Exuber<
000300 61 6e 74 20 43 74 61 67 73 2f 3b 22 09 65 78 74  >ant Ctags/;".ext<
000310 72 61 73 3a 70 73 65 75 64 6f 0d 0a 21 5f 54 41  >ras:pseudo..!_TA<
000320 47 5f 50 52 4f 47 52 41 4d 5f 55 52 4c 09 68 74  >G_PROGRAM_URL.ht<
000330 74 70 73 3a 2f 2f 63 74 61 67 73 2e 69 6f 2f 09  >tps://ctags.io/.<
000340 2f 6f 66 66 69 63 69 61 6c 20 73 69 74 65 2f 3b  >/official site/;<
000350 22 09 65 78 74 72 61 73 3a 70 73 65 75 64 6f 0d  >".extras:pseudo.<
000360 0a 21 5f 54 41 47 5f 50 52 4f 47 52 41 4d 5f 56  >.!_TAG_PROGRAM_V<
000370 45 52 53 49 4f 4e 09 35 2e 39 2e 30 09 2f 70 35  >ERSION.5.9.0./p5<
000380 2e 39 2e 32 30 32 32 30 34 32 34 2e 30 2f 3b 22  >.9.20220424.0/;"<
000390 09 65 78 74 72 61 73 3a 70 73 65 75 64 6f 0d 0a  >.extras:pseudo..<
0003a0 24 24 64 69 76 49 09 61 72 63 68 2f 70 61 72 69  >$$divI.arch/pari<
0003b0 73 63 2f 6b 65 72 6e 65 6c 2f 70 61 72 69 73 63  >sc/kernel/parisc<

$ od -A x -t x1z -v shorten.tags | head -60
000000 21 5f 43 49 54 52 45 5f 43 4d 44 09 63 74 61 67  >!_CITRE_CMD.ctag<
000010 73 7c 2d 6f 7c 25 54 41 47 53 46 49 4c 45 25 7c  >s|-o|%TAGSFILE%|<
000020 2d 2d 6c 61 6e 67 75 61 67 65 73 3d 43 2c 43 2b  >--languages=C,C+<
000030 2b 7c 2d 2d 6b 69 6e 64 73 2d 61 6c 6c 3d 2a 7c  >+|--kinds-all=*|<
000040 2d 2d 66 69 65 6c 64 73 3d 2a 7c 2d 2d 65 78 74  >--fields=*|--ext<
000050 72 61 73 3d 2a 7c 2d 52 09 2f 63 6f 6d 6d 61 6e  >ras=*|-R./comman<
000060 64 20 6c 69 6e 65 20 74 6f 20 67 65 6e 65 72 61  >d line to genera<
000070 74 65 20 74 68 69 73 20 74 61 67 73 20 66 69 6c  >te this tags fil<
000080 65 2f 3b 22 0d 0a 21 5f 54 41 47 5f 46 49 4c 45  >e/;"..!_TAG_FILE<
000090 5f 46 4f 52 4d 41 54 09 32 09 2f 65 78 74 65 6e  >_FORMAT.2./exten<
0000a0 64 65 64 20 66 6f 72 6d 61 74 3b 20 2d 2d 66 6f  >ded format; --fo<
0000b0 72 6d 61 74 3d 31 20 77 69 6c 6c 20 6e 6f 74 20  >rmat=1 will not <
0000c0 61 70 70 65 6e 64 20 3b 22 20 74 6f 20 6c 69 6e  >append ;" to lin<
0000d0 65 73 2f 3b 22 09 65 78 74 72 61 73 3a 70 73 65  >es/;".extras:pse<
0000e0 75 64 6f 0d 0a 21 5f 54 41 47 5f 46 49 4c 45 5f  >udo..!_TAG_FILE_<
0000f0 53 4f 52 54 45 44 09 31 09 2f 30 3d 75 6e 73 6f  >SORTED.1./0=unso<
000100 72 74 65 64 2c 20 31 3d 73 6f 72 74 65 64 2c 20  >rted, 1=sorted, <
000110 32 3d 66 6f 6c 64 63 61 73 65 2f 3b 22 09 65 78  >2=foldcase/;".ex<
000120 74 72 61 73 3a 70 73 65 75 64 6f 0d 0a 21 5f 54  >tras:pseudo..!_T<
000130 41 47 5f 4f 55 54 50 55 54 5f 45 58 43 4d 44 09  >AG_OUTPUT_EXCMD.<
000140 6d 69 78 65 64 09 2f 6e 75 6d 62 65 72 2c 20 70  >mixed./number, p<
000150 61 74 74 65 72 6e 2c 20 6d 69 78 65 64 2c 20 6f  >attern, mixed, o<
000160 72 20 63 6f 6d 62 69 6e 65 56 32 2f 3b 22 09 65  >r combineV2/;".e<
000170 78 74 72 61 73 3a 70 73 65 75 64 6f 0d 0a 21 5f  >xtras:pseudo..!_<
000180 54 41 47 5f 4f 55 54 50 55 54 5f 46 49 4c 45 53  >TAG_OUTPUT_FILES<
000190 45 50 09 73 6c 61 73 68 09 2f 73 6c 61 73 68 20  >EP.slash./slash <
0001a0 6f 72 20 62 61 63 6b 73 6c 61 73 68 2f 3b 22 09  >or backslash/;".<
0001b0 65 78 74 72 61 73 3a 70 73 65 75 64 6f 0d 0a 21  >extras:pseudo..!<
0001c0 5f 54 41 47 5f 4f 55 54 50 55 54 5f 4d 4f 44 45  >_TAG_OUTPUT_MODE<
0001d0 09 75 2d 63 74 61 67 73 09 2f 75 2d 63 74 61 67  >.u-ctags./u-ctag<
0001e0 73 20 6f 72 20 65 2d 63 74 61 67 73 2f 3b 22 09  >s or e-ctags/;".<
0001f0 65 78 74 72 61 73 3a 70 73 65 75 64 6f 0d 0a 21  >extras:pseudo..!<
000200 5f 54 41 47 5f 50 41 54 54 45 52 4e 5f 4c 45 4e  >_TAG_PATTERN_LEN<
000210 47 54 48 5f 4c 49 4d 49 54 09 39 36 09 2f 30 20  >GTH_LIMIT.96./0 <
000220 66 6f 72 20 6e 6f 20 6c 69 6d 69 74 2f 3b 22 09  >for no limit/;".<
000230 65 78 74 72 61 73 3a 70 73 65 75 64 6f 0d 0a 21  >extras:pseudo..!<
000240 5f 54 41 47 5f 50 52 4f 43 5f 43 57 44 09 44 3a  >_TAG_PROC_CWD.D:<
000250 2f 67 69 74 2f 6c 69 6e 75 78 2d 6d 61 73 74 65  >/git/linux-maste<
000260 72 5f 6d 6f 64 69 66 79 2f 6c 69 6e 75 78 2d 6d  >r_modify/linux-m<
000270 61 73 74 65 72 2f 09 2f 2f 3b 22 09 65 78 74 72  >aster/.//;".extr<
000280 61 73 3a 70 73 65 75 64 6f 0d 0a 21 5f 54 41 47  >as:pseudo..!_TAG<
000290 5f 50 52 4f 47 52 41 4d 5f 41 55 54 48 4f 52 09  >_PROGRAM_AUTHOR.<
0002a0 55 6e 69 76 65 72 73 61 6c 20 43 74 61 67 73 20  >Universal Ctags <
0002b0 54 65 61 6d 09 2f 2f 3b 22 09 65 78 74 72 61 73  >Team.//;".extras<
0002c0 3a 70 73 65 75 64 6f 0d 0a 21 5f 54 41 47 5f 50  >:pseudo..!_TAG_P<
0002d0 52 4f 47 52 41 4d 5f 4e 41 4d 45 09 55 6e 69 76  >ROGRAM_NAME.Univ<
0002e0 65 72 73 61 6c 20 43 74 61 67 73 09 2f 44 65 72  >ersal Ctags./Der<
0002f0 69 76 65 64 20 66 72 6f 6d 20 45 78 75 62 65 72  >ived from Exuber<
000300 61 6e 74 20 43 74 61 67 73 2f 3b 22 09 65 78 74  >ant Ctags/;".ext<
000310 72 61 73 3a 70 73 65 75 64 6f 0d 0a 21 5f 54 41  >ras:pseudo..!_TA<
000320 47 5f 50 52 4f 47 52 41 4d 5f 55 52 4c 09 68 74  >G_PROGRAM_URL.ht<
000330 74 70 73 3a 2f 2f 63 74 61 67 73 2e 69 6f 2f 09  >tps://ctags.io/.<
000340 2f 6f 66 66 69 63 69 61 6c 20 73 69 74 65 2f 3b  >/official site/;<
000350 22 09 65 78 74 72 61 73 3a 70 73 65 75 64 6f 0d  >".extras:pseudo.<
000360 0a 21 5f 54 41 47 5f 50 52 4f 47 52 41 4d 5f 56  >.!_TAG_PROGRAM_V<
000370 45 52 53 49 4f 4e 09 35 2e 39 2e 30 09 2f 70 35  >ERSION.5.9.0./p5<
000380 2e 39 2e 32 30 32 32 30 34 32 34 2e 30 2f 3b 22  >.9.20220424.0/;"<
000390 09 65 78 74 72 61 73 3a 70 73 65 75 64 6f 0d 0a  >.extras:pseudo..<
0003a0 24 24 64 69 76 49 09 61 72 63 68 2f 70 61 72 69  >$$divI.arch/pari<
0003b0 73 63 2f 6b 65 72 6e 65 6c 2f 70 61 72 69 73 63  >sc/kernel/parisc<
```

I have uploaded these two tags files on https://drive.google.com/file/d/1cMGlS9MItEgc_5Luc2P795Hdp1gqkZ_s/view?usp=sharing to accelerate the debugging process.

**Comment by masatake** [comment]
@lambertlulala, thank you. Surprisingly the od outputs are the same. 
Thank you for uploading the files. I received the file. As far as trying on Fedora35, readtags works well for both tags files.
I will try on mingw64.

**Comment by masatake** [comment]
Reproduced on mingw64.

**Comment by lambertlulala** [comment]
Thanks for your experiment!

**Comment by masatake** [comment]
I found the reason.
In libreadtags, a library used in the readtags command, ftell() is used to get the current file option in tags file.
ftell() returns `long` value. However, "long" is too small for the original.tags. original.tags is larger than 2GB.
https://stackoverflow.com/questions/4003405/32-bit-windows-and-the-2gb-file-size-limit-c-with-fseek-and-ftell

It will take time for fixing this.
Making the tags file smaller can be a workaround.

When running ctags, pass "--kinds-C=-zl" and "--kinds-C++=-zl" to the command line like:
```
ctags
-o
%TAGSFILE%
;; programming languages to be scanned
--languages=C,C++
--kinds-all=*
--kinds-C=-zl
--kinds-C++=-zl
--fields=*
--extras=*
-R
```

`z` means function parameters.
`l` means function local variables.
Not tagging these kinds may reduce the size of tags file.

**Comment by lambertlulala** [comment]
> I found the reason. In libreadtags, a library used in the readtags command, ftell() is used to get the current file position in tags file. ftell() returns `long` value. However, "long" is too small for the original.tags. original.tags is larger than 2GB. https://stackoverflow.com/questions/4003405/32-bit-windows-and-the-2gb-file-size-limit-c-with-fseek-and-ftell
> 
> It will take time for fixing this. Making the tags file smaller can be a workaround.
> 
> When running ctags, pass "--kinds-C=-zl" and "--kinds-C++=-zl" to the command line like:
> 
> ```
> ctags
> -o
> %TAGSFILE%
> ;; programming languages to be scanned
> --languages=C,C++
> --kinds-all=*
> --kinds-C=-zl
> --kinds-C++=-zl
> --fields=*
> --extras=*
> -R
> ```
> 
> `z` means function parameters. `l` means function local variables. Not tagging these kinds may reduce the size of tags file.

Adding these two arguments on my platform works without error. You are right, it's subtle, especially for the specific OS issue.

Thank you for your fix in advance! :smile:

**Comment by masatake** [comment]
I opened a new issue (https://github.com/universal-ctags/libreadtags/issues/36) for fixing the root cause.
ftell and fseek are used in ctags itself to manipulate a tags file. So just fixing libreadtags is not enough.

We need arithmetic operations about file positions, so fsetpos and fgetpos don't help.
https://stackoverflow.com/questions/4003405/32-bit-windows-and-the-2gb-file-size-limit-c-with-fseek-and-ftell

Anyway, I will close this.

**Comment by masatake** [comment]
@[lambertlulala](https://github.com/lambertlulala), BTW, if you use ctags for reading linux kernel, I recommend you to enable
`Kconfig` language like:
```
--languages=C,C++,Kconfig
```
`Asm` and `LdScript` can be enabled, too.
I expect enabling them doesn't make the tags output too large.

An alternative choice is running "make tags" at the top of linux kernel source tree.
However, I'm not sure whether the output tags file is suitable for citre.

**Comment by lambertlulala** [comment]
@masatake Thanks for handling this issue on mingw64. And these arguments for reading Linux kernel code are practical!

Normally the sizes of tags files generating from my daily working projects are less than 1GB. Besides, I think few people will read Linux kernel code on Windows since some filenames of source files conflict Windows, unless renaming them (it's annoying). Soon I will get a new laptop preloaded Ubuntu, then some issues occurred on Windows will never show up. :grinning:
